crates/holochain_zome_types/src/countersigning.rs

```diff
@@ -0,0 +1,288 @@
+    /// The session times.
+    /// Session headers must all have the same timestamp, which is the session offset.
+    session_times: CounterSigningSessionTimes,
+    /// The header information that is shared by all agents.
+    /// Contents depend on the header type, create, update, etc.
+    header_base: HeaderBase,
+    /// The preflight bytes for session.
+    preflight_bytes: PreflightBytes,
+}
+
+impl PreflightRequest {
+    /// Signing agents accessor.
+    pub fn signing_agents_ref(&self) -> &CounterSigningAgents {
+        &self.signing_agents
+    }
+
+    /// Enzyme index accessor.
+    pub fn enzyme_index_ref(&self) -> &Option<u8> {
+        &self.enzyme_index
+    }
+
+    /// Session times accessor.
+    pub fn session_times_ref(&self) -> &CounterSigningSessionTimes {
+        &self.session_times
+    }
+
+    /// Header base accessor.
+    pub fn header_base_ref(&self) -> &HeaderBase {
+        &self.header_base
+    }
+
+    /// Preflight bytes accessor.
+    pub fn preflight_bytes_ref(&self) -> &PreflightBytes {
+        &self.preflight_bytes
+    }
+}
+
+/// Every agent must send back a preflight response.
+/// All the preflight response data is signed by each agent and included in the session data.
+pub struct PreflightResponse {
+    /// The request this is a response to.
+    request: PreflightRequest,
+    /// The agent must provide their current chain state, state their position in the preflight and sign everything.
+    agent_state: CounterSigningAgentState,
+}
+
+impl PreflightResponse {
+    /// Request accessor.
+    pub fn request_ref(&self) -> &PreflightRequest {
+        &self.request
+    }
+
+    /// Agent state accessor.
+    pub fn agent_state_ref(&self) -> &CounterSigningAgentState {
+        &self.agent_state
+    }
+}
+
+/// Every countersigning agent must sign against their chain state.
+/// The chain must be frozen until each agent decides to sign or exit the session.
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct CounterSigningAgentState {
+    /// The index of the agent in the preflight request agent vector.
+    agent_index: u8,
+    /// The current (frozen) top of the agent's local chain.
+    chain_top: HeaderHash,
+    /// The header sequence of the agent's chain top.
+    header_seq: u32,
+    /// The signature of all preflight request data and agent state data.
+    signature: Signature,
+}
+
+impl CounterSigningAgentState {
+    /// Agent index accessor.
+    pub fn agent_index(&self) -> u8 {
+        self.agent_index
+    }
+
+    /// Chain top accessor.
+    pub fn chain_top_ref(&self) -> &HeaderHash {
+        &self.chain_top
+    }
+
+    /// Header seq accessor.
+    pub fn header_seq(&self) -> u32 {
+        self.header_seq
+    }
+
+    /// Signature accessor.
+    pub fn signature_ref(&self) -> &Signature {
+        &self.signature
+    }
+}
+
+/// A vector of agents to countersign a shared entry.
+/// The vector must be sorted to generate a CounterSigningTag.
```

Comment:
Should `CounterSigningAgentState` impl `Ord` then?

Reply:
ah! this is an old requirement, the comment is cruft so i deleted it instead
now all that matters is that whatever order the agents are listed in they know their own index and sign the same vec